### ISSUE TITLE
YAML-parameters won't read correctly for two or more composable nodes with non-empty root namespace

### ISSUE THREAD

**Issue by AlexeyMerzlyakov** [ros2/launch_ros#344]
When running two or more component nodes inside one container with some non-empty root namespace, composable nodes inside this container can not read ROS-parameters from YAML correctly.

## Bug report

**Required Info:**

- Operating System:
  - Ubuntu 20.04, 22.04
- Installation type:
  - ROS2 Rolling built from sources (check-up date 2023.01.10)
- Version or commit hash:
  - 35b2ca9
- DDS implementation:
  - CycloneDDS
- Client library (if applicable):
  - N/A

#### Steps to reproduce issue
Please, consider the attached example package: [test_pkg.zip](https://github.com/ros2/launch_ros/files/10468727/test_pkg.zip)
This example contains two nodes, exported as composable ones, loaded into a running container. Both nodes are running in some non-empty root namespace `robot1`. Each node is reading `node_param` ROS-parameter from input YAML-file (by-default parameter's value is `UNDEFINED`):
```
robot1:
  node1:
    ros__parameters:
      node_param: "PASS"
  node2:
    ros__parameters:
      node_param: "PASS"
```

Build test package and launch it as follows:
```
ros2 launch test_pkg test_launch.py use_composition:=True
```

#### Expected behavior
Both parameters to be set to `PASS` value, that could be observed from logs.

#### Actual behavior
For the case of non-empty `robot1` root namespace and we could observe some kind of undefined behavior: for the first initialized node this parameter to be read from input YAML, while for other ones it can not and to be left to default:
```
[amerzlyakov@amerzlyakov-PC test_ws]$ ros2 launch test_pkg test_launch.py use_composition:=True
[INFO] [launch]: All log files can be found below /home/amerzlyakov/.ros/log/2023-01-20-20-20-50-447365-amerzlyakov-PC-102330
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [component_container_isolated-1]: process started with pid [102344]
[component_container_isolated-1] [INFO] [1674235250.925593065] [robot1.test_container]: Load Library: /home/amerzlyakov/ROS2/2022/2022.09.02_multi_robot_issue/issue1_2_ros_params_composable_nodes/20230120_4_test_simplification/test_ws/install/test_pkg/lib/libnode1_core.so
[component_container_isolated-1] [INFO] [1674235250.926370893] [robot1.test_container]: Found class: rclcpp_components::NodeFactoryTemplate<test_pkg::Node1>
[component_container_isolated-1] [INFO] [1674235250.926397855] [robot1.test_container]: Instantiate class: rclcpp_components::NodeFactoryTemplate<test_pkg::Node1>
[component_container_isolated-1] [INFO] [1674235250.932592006] [robot1.node1]: NODE1 name = node1, namespace = /robot1
[component_container_isolated-1] [INFO] [1674235250.932630453] [robot1.node1]: node_param = PASS
[INFO] [launch_ros.actions.load_composable_nodes]: Loaded node '/robot1/node1' in container 'robot1/test_container'
[component_container_isolated-1] [INFO] [1674235250.937744616] [robot1.test_container]: Load Library: /home/amerzlyakov/ROS2/2022/2022.09.02_multi_robot_issue/issue1_2_ros_params_composable_nodes/20230120_4_test_simplification/test_ws/install/test_pkg/lib/libnode2_core.so
[component_container_isolated-1] [INFO] [1674235250.938156367] [robot1.test_container]: Found class: rclcpp_components::NodeFactoryTemplate<test_pkg::Node2>
[component_container_isolated-1] [INFO] [1674235250.938182809] [robot1.test_container]: Instantiate class: rclcpp_components::NodeFactoryTemplate<test_pkg::Node2>
[component_container_isolated-1] [INFO] [1674235250.943180176] [robot1.node2]: NODE2 name = node2, namespace = /robot1
[component_container_isolated-1] [INFO] [1674235250.943213352] [robot1.node2]: node_param = UNDEFINED
[INFO] [launch_ros.actions.load_composable_nodes]: Loaded node '/robot1/node2' in container 'robot1/test_container'
```

Running the same example, but with conventional separate nodes will give correct result:
```
[amerzlyakov@amerzlyakov-PC test_ws]$ ros2 launch test_pkg test_launch.py use_composition:=False
[INFO] [launch]: All log files can be found below /home/amerzlyakov/.ros/log/2023-01-20-20-20-45-051836-amerzlyakov-PC-102302
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [node1-1]: process started with pid [102303]
[INFO] [node2-2]: process started with pid [102305]
[node2-2] [INFO] [1674235245.273683472] [robot1.node2]: NODE2 name = node2, namespace = /robot1
[node2-2] [INFO] [1674235245.273809217] [robot1.node2]: node_param = PASS
[node1-1] [INFO] [1674235245.273741356] [robot1.node1]: NODE1 name = node1, namespace = /robot1
[node1-1] [INFO] [1674235245.273822142] [robot1.node1]: node_param = PASS
```

#### Additional information
When using no root namespace, everything works correctly, for both conventional and composable cases.
The fact that undefined behavior appears only for composable nodes with non-empty root namespace leads to thought that the problem might be in composable node running mechanism, but it should be checked.

**Comment by AlexeyMerzlyakov** [comment]
The problem was found in  `normalize_parameters_dict()` routine, which is not keeping `keys` correctly, that making wrong `normalized_param_dict` for the example from this issue:
```
normalized_param_dict =  {'robot1/node1': {'node_param': 'PASS1'}, 'node2': {'node_param': 'PASS2'}}
```
The `normalized_param_dict` should contain all node names to be full in order to cause `to_parameters_list()` to operate correctly:
```
normalized_param_dict =  {'robot1/node1': {'node_param': 'PASS1'}, 'robot1/node2': {'node_param': 'PASS2'}}
```

This seems to appear because of too strong cut-off of the `keys` list in the  `normalize_parameters_dict()` routine:
```
    def normalize_parameters_dict(dictionary, keys, result_dict):
            ...
            if isinstance(value, dict):
                keys.append(key.lstrip('/'))
                result_dict = normalize_parameters_dict(value, keys, result_dict)
                # Reset keys in case there are multiple ros__parameter entries
                keys = []
```

The latest string should pop only the latest `keys` value after the returning back from the recursion, instead of full cleanup:
```
                result_dict = normalize_parameters_dict(value, keys, result_dict)
                # Reset keys in case there are multiple ros__parameter entries
                keys.pop()
```
... which fixes situation.

**Comment by AlexeyMerzlyakov** [comment]
Proposed fix #347
